File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.ts
function minFlips(s: string): number {
    //optimized
    let n:number=s.length;

    let ans:number=Infinity;

    let flip1:number=0;
    let flip2:number=0;

    let i:number=0;
    let j:number=0;

    while(j<2*n){
        let expChar1=j%2?'1':'0';
        let expChar2=j%2?'0':'1';

        if(s[j%n]!==expChar1) flip1++;

        if(s[j%n]!==expChar2) flip2++;

        if(j-i+1>n){
            expChar1=i%2?'1':'0';
            expChar2=i%2?'0':'1';

            if(s[i%n]!==expChar1) flip1--;
            if(s[i%n]!==expChar2) flip2--;
            
            i++;
        }

        if(j-i+1===n){
            ans=Math.min(ans,flip1,flip2);
        }

        j++;
    }

    return ans;
};

File: 1888-minimum-number-of-flips-to-make-the-binary-string-alternating/1888-minimum-number-of-flips-to-make-the-binary-string-alternating.cpp
class Solution {
public:
    int minFlips(string s) {
        int n = s.length();
        string s1,s2;

        for(int i=0;i<2*n;i++){
            s1+=(i%2?'1':'0');
            s2+=(i%2?'0':'1');
        }

        int flip1=0;
        int flip2=0;

        int ans=INT_MAX;

        int j=0;
        int i=0;

        while(j<2*n){
            if(s[j%n]!=s1[j]) flip1++;
                
            if(s[j%n]!=s2[j]) flip2++;

            if(j-i+1>n){
                if(s[i%n]!=s1[i]) flip1--;
                if(s[i%n]!=s2[i]) flip2--;
                i++;
            }

            if(j-i+1==n){
                //only consider updating the size if window's size is equal to the string's size
                ans=min({ans,flip1,flip2});
            }
            j++;
        }

        return ans;
    }
};